### PR TITLE
WIP: Add support for calling AddAutoMapper multiple times

### PR DESCRIPTION
**Pull request by prochnowc**
- Introduced MapperBuilder which adds support for calling AddAutoMapper() multiple times.
- Refactored existing ServiceCollectionExtensions (split out extension methods and CandidateResolver)

### CODE REVIEW

File: test/AutoMapper.Extensions.Microsoft.DependencyInjection.Tests/Profiles.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public Profile1()
 
     public abstract class AbstractProfile : Profile { }
 
-    internal class Profile2 : Profile
+    public class Profile2 : Profile
```
Comment (prochnowc) [inline test/AutoMapper.Extensions.Microsoft.DependencyInjection.Tests/Profiles.cs]: I made this public because `AddAutoMapper()` had different behavior compared to `IConfigurationExpression.AddProfiles()`. `AddProfiles()` only adds public Profiles.

**Comment by jbogard** [comment]
I like the idea - I'm wondering what if anything could be pushed to the core library to make this easier. The primary reason we just have `Action<IMapperConfigurationExpression>` is because that's what the API accepts. I added `AddProfiles` in part because of this.

The other thing I would look at would instead of adding overloads that take MapperBuilder and the other parameters, just remove those in favor of adding MapperBuilder methods that capture that information and apply it during configuration. That way you can either go 100% MapperBuilder route, or the simple version.

**Comment by prochnowc** [comment]
The way i built `MapperBuilder` makes the implementation independent of the `IMapperConfigurationExpression` - The interface can be modified in the Core project but the implementation for Extensions.DependencyInjection doesnt need modification.

Also i did the approach with extension methods because they all depend on single `UseConfiguration` method (which makes `IMapperBuilder` interface very stable), again independently of the signature of the `IMapperConfigurationExpression` interface.
Having `MapperBuilder` methods capture the info and later apply them will make the implementation far more complex.

The simple way can be added with overloads of the UseAutoMapper method which do not take any parameters - they will internally call apropriate MapperBuilder methods.

**Comment by prochnowc** [comment]
@jbogard if you are willing to eventually push this to next major release of automapper.extensions.dependencyinjection i can refactor the servicecollection extensions without keeping breaking changes in mind.

**Comment by alexinea** [comment]
I like the idea (・ω<) ﾃﾍﾍﾟﾛ and I need it, too.

**Comment by prochnowc** [comment]
@jbogard So should i keep breaking changes in mind? Or do you intent to release the changes (if approved) as next major version of AutoMapper.Extensions.Microsoft.DependencyInjection ?

What about my other comments?

**Comment by jbogard** [comment]
The other thing I'm looking at is getting rid of the DependencyContext stuff, and forcing you to explicitly specify the assemblies to use.

Unfortunately this is also a breaking change too.

**Comment by VictorioBerra** [comment]
Where do we stand with this? This issue has came up for me again because I now use XUnit with an IFixture and a custom WebApplicationFactory. So if i have two classes I get the dreaded "mapper already initialized" because the WebApplicationFactory will configure services again for my second class. I still have not figured out how or why.

**Comment by lbargaoanu** [comment]
@VictorioBerra https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection/search?utf8=%E2%9C%93&q=UseStaticRegistration+&type=

**Comment by VictorioBerra** [comment]
Setting `AutoMapper.ServiceCollectionExtensions.UseStaticRegistration = false;` still did not help me. Ill create a new issue with a sample application.

**Comment by lbargaoanu** [comment]
Maybe, but that's a problem with your own code, not with AM. The error happens only in the static Mapper.

**Comment by VictorioBerra** [comment]
@lbargaoanu 

If i understand correctly, using `.AddAutoMapper()` means you will get a static configuration, and IMapper instance scoped to a request.

This means that if anyone is using `.AddAutoMapper()` and they also want to follow the official Microsoft tutorial and example for writing tests using the new WebApplicationFactory as shown [here](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1) then their tests will fail with the error:  `Message: System.InvalidOperationException : Mapper already initialized. You must call Initialize once per application domain/process.`

I have a sample app I created using the absolute bare minimum code from basically `dotnet new mvc && dotnet new xunit` where I AddAutoMapper and I also have two tests class using a pretty much empty WebApplicationFactory. I also tossed a single profile in the mvc app. This was incredibly easy to reproduce. and `Mapper.Reset();` in my startup did not help, as well as disabling `UseStaticRegistration`. 

I have uploaded the code and will reference this here in just a second.
